src/omv/img/draw.c

```diff
@@ -474,55 +488,152 @@ static int safe_map_pixel(int dst_bpp, int src_bpp, int pixel)
-                        int b = COLOR_RGB565_TO_B5(other_pixel)*alpha + COLOR_RGB565_TO_B5(img_pixel)*neg_alpha;
-                        result_pixel = COLOR_R5_G6_B5_TO_RGB565(r, g, b);
-                        break;
+                }
+            }
+            break;
+        }
+        case IMAGE_BPP_GRAYSCALE: {
+            // Iterate the img area to be updated
+            for (int y = other_y_start; y < other_y_end; y++) {
+                // Pre-add x_off here to save adding it inside the central loop
+                uint8_t *img_row_ptr = IMAGE_COMPUTE_GRAYSCALE_PIXEL_ROW_PTR(img, y_off + y) + x_off;
+
+                const int other_y = fast_floorf(y * over_yscale);
+                void *other_row_ptr = imlib_compute_row_ptr(other, other_y);
+
+                for (int x = other_x_start; x < other_x_end; x++) {
+                    const int other_x = fast_floorf(x * over_xscale);
+
+                    if ((!mask) || image_get_mask_pixel(mask, other_x, other_y)) {
+                        const uint8_t other_pixel = safe_map_pixel(IMAGE_BPP_GRAYSCALE, other_bpp, imlib_get_pixel_fast(other_bpp, other_row_ptr, other_x));
+
+                        uint8_t result_pixel;
+                        if (alpha==256) {
+                            result_pixel = other_pixel;
+                        }
+                        else {
+                            const uint8_t img_pixel = IMAGE_GET_GRAYSCALE_PIXEL_FAST(img_row_ptr,  x);
+                            
+                            const uint32_t vgs = (other_pixel << 16) + img_pixel;
+                            result_pixel = __SMUAD(va, vgs)>>8;
+                        }
+
+                        IMAGE_PUT_GRAYSCALE_PIXEL_FAST(img_row_ptr, x, result_pixel);
                     }
-                    default: {
-                        result_pixel = 0;
-                        break;
+                }
+            }
+            break;
+        }
+        case IMAGE_BPP_RGB565: {
+            // Iterate the img area to be updated
+            for (int y = other_y_start; y < other_y_end; y++) {
+                // Pre-add x_off here to save adding it inside the central loop
+                uint16_t *img_row_ptr = IMAGE_COMPUTE_RGB565_PIXEL_ROW_PTR(img, y_off + y) + x_off;
+                
+                const int other_y = fast_floorf(y * over_yscale);
+                void *other_row_ptr = imlib_compute_row_ptr(other, other_y);
+
+                for (int x = other_x_start; x < other_x_end; x++) {
+                    const int other_x = fast_floorf(x * over_xscale);
+
+                    if ((!mask) || image_get_mask_pixel(mask, other_x, other_y)) {
+                        uint32_t other_pixel = imlib_get_pixel_fast(other_bpp, other_row_ptr, other_x);
+                        other_pixel = color_palette
+                            ? color_palette[other_pixel]
```

Comment:
When a color_palette is specified other_pixel can only be grayscale.  There's a check in py_image_draw_image to prevent color_palettes with anything other than grayscale type other images.